src/lib.rs

```diff
@@ -128,6 +130,7 @@ where
         } else {
             self.dio.set_low()?;
         }
+        self.delay();
```

Comment:
Could you please move this before `if` and test whether it works or not? It would make it more symmetric, e.g.:
- low CLK, delay
- set DIO
- high CLK, delay
- (sometimes in `start` and `stop` routines) set DIO

Reply:
Yes, that sounds like a good idea. The current patch is intentionally doing the minimal change to make the clock pulse longer if necessary.
Do you have a better data sheet than https://raw.githubusercontent.com/avishorp/TM1637/master/docs/TM1637_V2.4_EN.pdf? The timing information in that one confuses me a bit. It doesn't help that the table on the bottom of page 10 lists some minimum timings without exactly telling what they measure.

Reply:
Wait: No! With that change, now there is no longer a delay between two bits:
- first bit:
  - low CLK
  - delay
  - set DIO
  - delay
  - high CLK
- next bit:
  - low CLK
  - ...
  
So there's the same problem as initially, only that now the high time between two bits can become arbitrarily short. That doesn't work reliably either.

Reply:
I also tested this variant, it works but is less reliable:
```
      fn send_bit_and_delay(&mut self, value: Bit) -> Res<E> {
          self.clk.set_low()?;
          self.delay();
          if let Bit::ONE = value {
              self.dio.set_high()?;
          } else {
              self.dio.set_low()?;
          }
          self.clk.set_high()?;
          self.delay();
  
          Ok(())
      }
```
In this combination, if I use a push-pull pin for CLK, it does no longer work.

I think the issue with this is that the DIO change happens to shortly before the rising edge of CLK. If the DIO line rises slightly slower than the CLK line, the data line will be sampled before the signal level is valid.
(This could also happen if both CLK and DIO are open-drain outputs, if the rise times are different due to tolerances in the pull-up resistors.)

In summary, this version seems to be the most reliable one, at least without adding additional delay statements:
```
    fn send_bit_and_delay(&mut self, value: Bit) -> Res<E> {
        self.clk.set_low()?;
        if let Bit::ONE = value {
            self.dio.set_high()?;
        } else {
            self.dio.set_low()?;
        }
        self.delay();
        self.clk.set_high()?;
        self.delay();

        Ok(())
    }
```

Another option would be to half the individual delays again, and have delays both before and after setting DIO. But that would require delays shorther than 1µs, or would slow down the communication.

Reply:
> Wait: No! With that change, now there is no longer a delay between two bits:

But lib has delay just after set_high - https://github.com/igelbox/tm1637-rs/blob/25e5e7e72ed7eed2e18d40db8664b355b677d76c/src/lib.rs#L135

> In this combination, if I use a push-pull pin for CLK, it does no longer work.

Okay. I have no hardware to test now. So, I'll choose the variant that works for you.